File: csharp/NatsHttpGateway/Examples/websocket_client_example.py
#!/usr/bin/env python3
"""
Example Python WebSocket client for NatsHttpGateway streaming endpoints.

Requirements:
    pip install protobuf websockets

Usage:
    python websocket_client_example.py
"""

import sys
import os
import asyncio
import websockets
from datetime import datetime

# Import generated protobuf classes
try:
    import message_pb2
except ImportError:
    print("Error: message_pb2 not found.")
    print("Generate it with: protoc --python_out=. ../Protos/message.proto")
    sys.exit(1)


class WebSocketClient:
    """Client for streaming messages via WebSocket from NatsHttpGateway"""

    def __init__(self, base_url="http://localhost:5000"):
        self.base_url = base_url
        self.ws_base_url = base_url.replace("http://", "ws://").replace("https://", "wss://")

    async def stream_from_ephemeral_consumer(self, subject_filter="events.>", max_messages=10):
        """Example 1: Stream messages from an ephemeral consumer using subject filter"""
        print(f"=== Example 1: Streaming from Ephemeral Consumer ({subject_filter}) ===")

        ws_url = f"{self.ws_base_url}/ws/websocketmessages/{subject_filter}"
        print(f"Connecting to: {ws_url}")

        try:
            async with websockets.connect(ws_url) as websocket:
                print("✓ WebSocket connected")

                message_count = 0

                while message_count < max_messages:
                    try:
                        # Receive binary protobuf frame
                        frame_bytes = await websocket.recv()

                        # Parse the WebSocketFrame
                        frame = message_pb2.WebSocketFrame()
                        frame.ParseFromString(frame_bytes)

                        if frame.type == message_pb2.CONTROL:
                            self._handle_control_message(frame.control)
                        elif frame.type == message_pb2.MESSAGE:
                            self._handle_stream_message(frame.message)
                            message_count += 1

                    except websockets.exceptions.ConnectionClosed as e:
                        print(f"✓ Connection closed: {e}")
                        break

                print(f"✓ Received {message_count} messages")

        except Exception as e:
            print(f"✗ Error: {e}")

        print()

    async def stream_from_durable_consumer(
        self,
        stream_name="EVENTS",
        consumer_name="my-durable-consumer",
        max_messages=10
    ):
        """Example 2: Stream messages from a durable consumer"""
        print(f"=== Example 2: Streaming from Durable Consumer ({consumer_name}) ===")

        ws_url = f"{self.ws_base_url}/ws/websocketmessages/{stream_name}/consumer/{consumer_name}"
        print(f"Connecting to: {ws_url}")

        try:
            async with websockets.connect(ws_url) as websocket:
                print("✓ WebSocket connected")

                message_count = 0

                while message_count < max_messages:
                    try:
                        # Receive binary protobuf frame
                        frame_bytes = await websocket.recv()

                        # Parse the WebSocketFrame
                        frame = message_pb2.WebSocketFrame()
                        frame.ParseFromString(frame_bytes)

                        if frame.type == message_pb2.CONTROL:
                            self._handle_control_message(frame.control)
                        elif frame.type == message_pb2.MESSAGE:
                            self._handle_stream_message(frame.message)
                            message_count += 1

                    except websockets.exceptions.ConnectionClosed as e:
                        print(f"✓ Connection closed: {e}")
                        break

                print(f"✓ Received {message_count} messages")

        except websockets.exceptions.InvalidStatusCode as e:
            print(f"✗ Error: {e}")
            print(f"  Make sure the durable consumer '{consumer_name}' exists in stream '{stream_name}'")
        except Exception as e:
            print(f"✗ Error: {e}")

        print()

    async def stream_with_timeout(self, subject_filter="events.test", timeout_seconds=30):
        """Example 3: Stream with custom message processing and timeout"""
        print(f"=== Example 3: Streaming with Timeout ({timeout_seconds}s) ===")

        ws_url = f"{self.ws_base_url}/ws/websocketmessages/{subject_filter}"
        print(f"Connecting to: {ws_url}")

        try:
            async with websockets.connect(ws_url) as websocket:
                print("✓ WebSocket connected")
                print(f"  Will disconnect after {timeout_seconds} seconds")

                message_count = 0
                start_time = datetime.now()

                try:
                    while True:
                        # Receive with timeout
                        frame_bytes = await asyncio.wait_for(
                            websocket.recv(),
                            timeout=timeout_seconds
                        )

                        # Parse the WebSocketFrame
                        frame = message_pb2.WebSocketFrame()
                        frame.ParseFromString(frame_bytes)

                        if frame.type == message_pb2.CONTROL:
                            self._handle_control_message(frame.control)
                        elif frame.type == message_pb2.MESSAGE:
                            message_count += 1
                            elapsed = datetime.now() - start_time
                            minutes, seconds = divmod(elapsed.total_seconds(), 60)
                            print(f"  [{int(minutes):02d}:{int(seconds):02d}] Message #{message_count}: "
                                  f"{frame.message.subject} (seq: {frame.message.sequence})")

                except asyncio.TimeoutError:
                    print(f"✓ Timeout reached after {timeout_seconds} seconds")

                except websockets.exceptions.ConnectionClosed as e:
                    print(f"✓ Connection closed: {e}")

                print(f"✓ Stream ended - received {message_count} messages")

        except Exception as e:
            print(f"✗ Error: {e}")

        print()

    def _handle_control_message(self, control):
        """Handle control messages from the server"""
        icons = {
            message_pb2.ERROR: "✗",
            message_pb2.SUBSCRIBE_ACK: "✓",
            message_pb2.CLOSE: "✓",
            message_pb2.KEEPALIVE: "♥"
        }
        icon = icons.get(control.type, "•")

        type_name = message_pb2.ControlType.Name(control.type)
        print(f"{icon} Control [{type_name}]: {control.message}")

    def _handle_stream_message(self, message):
        """Handle stream messages from NATS"""
        print("  Message received:")
        print(f"    Subject:  {message.subject}")
        print(f"    Sequence: {message.sequence}")
        print(f"    Size:     {message.size_bytes} bytes")

        if message.HasField('timestamp'):
            timestamp = datetime.fromtimestamp(message.timestamp.seconds)
            millis = message.timestamp.nanos // 1000000
            print(f"    Time:     {timestamp.strftime('%Y-%m-%d %H:%M:%S')}.{millis:03d}")

        if message.consumer:
            print(f"    Consumer: {message.consumer}")

        # Try to decode data as UTF-8 string
        if message.data:
            try:
                data_str = message.data.decode('utf-8')
                preview = data_str[:100] + "..." if len(data_str) > 100 else data_str
                print(f"    Data:     {preview}")
            except:
                print(f"    Data:     [binary, {len(message.data)} bytes]")

        print()

    async def run_all_examples(self):
        """Run all examples"""
        print(f"WebSocket Python Client - Connecting to {self.base_url}")
        print("=" * 80)
        print()

        try:
            # Example 1: Stream from ephemeral consumer
            await self.stream_from_ephemeral_consumer("events.>", max_messages=5)

            # Example 2: Stream from durable consumer (will fail if consumer doesn't exist)
            # Uncomment if you have a durable consumer set up:
            # await self.stream_from_durable_consumer("EVENTS", "my-durable-consumer", max_messages=5)

            # Example 3: Stream with timeout
            await self.stream_with_timeout("events.test", timeout_seconds=10)

            print("=" * 80)
            print("✓ All examples completed successfully!")

        except Exception as e:
            print(f"✗ Error: {e}")
            print(f"  Make sure NatsHttpGateway is running at {self.base_url}")
            print("  Make sure NATS is running and has messages to stream")


async def main():
    """Main entry point"""
    # Configuration priority: CLI arg > Environment variable > Default
    base_url = (
        sys.argv[1] if len(sys.argv) > 1
        else os.getenv("NATS_GATEWAY_URL", "http://localhost:5000")
    )
    client = WebSocketClient(base_url)
    await client.run_all_examples()


if __name__ == "__main__":
    asyncio.run(main())

File: csharp/NatsHttpGateway/Examples/WebSocketClientExample.cs
using NatsHttpGateway.Protos;
using Google.Protobuf;
using System.Net.WebSockets;
using System.Text;

namespace NatsHttpGateway.Examples;

/// <summary>
/// Example client demonstrating how to use WebSocket streaming endpoints
/// </summary>
public class WebSocketClientExample
{
    private readonly string _baseUrl;
    private readonly string _wsBaseUrl;

    public WebSocketClientExample(string? baseUrl = null)
    {
        // Configuration priority: Parameter > Environment variable > Default
        _baseUrl = baseUrl
            ?? Environment.GetEnvironmentVariable("NATS_GATEWAY_URL")
            ?? "http://localhost:5000";
        _wsBaseUrl = _baseUrl.Replace("http://", "ws://").Replace("https://", "wss://");
    }

    /// <summary>
    /// Example 1: Stream messages from an ephemeral consumer using subject filter
    /// </summary>
    public async Task StreamFromEphemeralConsumer(string subjectFilter = "events.>", int maxMessages = 10)
    {
        Console.WriteLine($"=== Example 1: Streaming from Ephemeral Consumer ({subjectFilter}) ===");

        var wsUrl = $"{_wsBaseUrl}/ws/websocketmessages/{subjectFilter}";
        Console.WriteLine($"Connecting to: {wsUrl}");

        using var ws = new ClientWebSocket();
        var cts = new CancellationTokenSource();

        try
        {
            await ws.ConnectAsync(new Uri(wsUrl), cts.Token);
            Console.WriteLine("✓ WebSocket connected");

            var messageCount = 0;

            while (ws.State == WebSocketState.Open && messageCount < maxMessages)
            {
                var buffer = new byte[1024 * 16]; // 16KB buffer
                var result = await ws.ReceiveAsync(new ArraySegment<byte>(buffer), cts.Token);

                if (result.MessageType == WebSocketMessageType.Binary)
                {
                    var frameBytes = new byte[result.Count];
                    Array.Copy(buffer, frameBytes, result.Count);

                    // Parse the WebSocketFrame
                    var frame = WebSocketFrame.Parser.ParseFrom(frameBytes);

                    switch (frame.Type)
                    {
                        case FrameType.Control:
                            HandleControlMessage(frame.Control);
                            break;

                        case FrameType.Message:
                            HandleStreamMessage(frame.Message);
                            messageCount++;
                            break;
                    }
                }
                else if (result.MessageType == WebSocketMessageType.Close)
                {
                    Console.WriteLine($"✓ Server closed connection: {result.CloseStatusDescription}");
                    break;
                }
            }

            Console.WriteLine($"✓ Received {messageCount} messages");

            // Close the WebSocket gracefully
            if (ws.State == WebSocketState.Open)
            {
                await ws.CloseAsync(WebSocketCloseStatus.NormalClosure, "Client closing", CancellationToken.None);
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"✗ Error: {ex.Message}");
        }
        finally
        {
            cts.Cancel();
        }

        Console.WriteLine();
    }

    /// <summary>
    /// Example 2: Stream messages from a durable consumer
    /// </summary>
    public async Task StreamFromDurableConsumer(
        string streamName = "EVENTS",
        string consumerName = "my-durable-consumer",
        int maxMessages = 10)
    {
        Console.WriteLine($"=== Example 2: Streaming from Durable Consumer ({consumerName}) ===");

        var wsUrl = $"{_wsBaseUrl}/ws/websocketmessages/{streamName}/consumer/{consumerName}";
        Console.WriteLine($"Connecting to: {wsUrl}");

        using var ws = new ClientWebSocket();
        var cts = new CancellationTokenSource();

        try
        {
            await ws.ConnectAsync(new Uri(wsUrl), cts.Token);
            Console.WriteLine("✓ WebSocket connected");

            var messageCount = 0;

            while (ws.State == WebSocketState.Open && messageCount < maxMessages)
            {
                var buffer = new byte[1024 * 16]; // 16KB buffer
                var result = await ws.ReceiveAsync(new ArraySegment<byte>(buffer), cts.Token);

                if (result.MessageType == WebSocketMessageType.Binary)
                {
                    var frameBytes = new byte[result.Count];
                    Array.Copy(buffer, frameBytes, result.Count);

                    // Parse the WebSocketFrame
                    var frame = WebSocketFrame.Parser.ParseFrom(frameBytes);

                    switch (frame.Type)
                    {
                        case FrameType.Control:
                            HandleControlMessage(frame.Control);
                            break;

                        case FrameType.Message:
                            HandleStreamMessage(frame.Message);
                            messageCount++;
                            break;
                    }
                }
                else if (result.MessageType == WebSocketMessageType.Close)
                {
                    Console.WriteLine($"✓ Server closed connection: {result.CloseStatusDescription}");
                    break;
                }
            }

            Console.WriteLine($"✓ Received {messageCount} messages");

            // Close the WebSocket gracefully
            if (ws.State == WebSocketState.Open)
            {
                await ws.CloseAsync(WebSocketCloseStatus.NormalClosure, "Client closing", CancellationToken.None);
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"✗ Error: {ex.Message}");
            Console.WriteLine($"  Make sure the durable consumer '{consumerName}' exists in stream '{streamName}'");
        }
        finally
        {
            cts.Cancel();
        }

        Console.WriteLine();
    }

    /// <summary>
    /// Example 3: Stream with custom message processing and timeout
    /// </summary>
    public async Task StreamWithTimeout(string subjectFilter = "events.test", int timeoutSeconds = 30)
    {
        Console.WriteLine($"=== Example 3: Streaming with Timeout ({timeoutSeconds}s) ===");

        var wsUrl = $"{_wsBaseUrl}/ws/websocketmessages/{subjectFilter}";
        Console.WriteLine($"Connecting to: {wsUrl}");

        using var ws = new ClientWebSocket();
        var cts = new CancellationTokenSource(TimeSpan.FromSeconds(timeoutSeconds));

        try
        {
            await ws.ConnectAsync(new Uri(wsUrl), cts.Token);
            Console.WriteLine("✓ WebSocket connected");
            Console.WriteLine($"  Will disconnect after {timeoutSeconds} seconds or when cancelled");

            var messageCount = 0;
            var startTime = DateTime.UtcNow;

            while (ws.State == WebSocketState.Open)
            {
                var buffer = new byte[1024 * 16];
                var result = await ws.ReceiveAsync(new ArraySegment<byte>(buffer), cts.Token);

                if (result.MessageType == WebSocketMessageType.Binary)
                {
                    var frameBytes = new byte[result.Count];
                    Array.Copy(buffer, frameBytes, result.Count);

                    var frame = WebSocketFrame.Parser.ParseFrom(frameBytes);

                    switch (frame.Type)
                    {
                        case FrameType.Control:
                            HandleControlMessage(frame.Control);
                            break;

                        case FrameType.Message:
                            messageCount++;
                            var elapsed = DateTime.UtcNow - startTime;
                            Console.WriteLine($"  [{elapsed:mm\\:ss}] Message #{messageCount}: {frame.Message.Subject} (seq: {frame.Message.Sequence})");
                            break;
                    }
                }
                else if (result.MessageType == WebSocketMessageType.Close)
                {
                    Console.WriteLine($"✓ Server closed connection: {result.CloseStatusDescription}");
                    break;
                }
            }

            Console.WriteLine($"✓ Stream ended - received {messageCount} messages");
        }
        catch (OperationCanceledException)
        {
            Console.WriteLine($"✓ Timeout reached after {timeoutSeconds} seconds");
        }
        catch (Exception ex)
        {
            Console.WriteLine($"✗ Error: {ex.Message}");
        }
        finally
        {
            if (ws.State == WebSocketState.Open)
            {
                await ws.CloseAsync(WebSocketCloseStatus.NormalClosure, "Timeout", CancellationToken.None);
            }
        }

        Console.WriteLine();
    }

    /// <summary>
    /// Handle control messages from the server
    /// </summary>
    private void HandleControlMessage(ControlMessage control)
    {
        var icon = control.Type switch
        {
            ControlType.Error => "✗",
            ControlType.SubscribeAck => "✓",
            ControlType.Close => "✓",
            ControlType.Keepalive => "♥",
            _ => "•"
        };

        Console.WriteLine($"{icon} Control [{control.Type}]: {control.Message}");
    }

    /// <summary>
    /// Handle stream messages from NATS
    /// </summary>
    private void HandleStreamMessage(StreamMessage message)
    {
        Console.WriteLine($"  Message received:");
        Console.WriteLine($"    Subject:  {message.Subject}");
        Console.WriteLine($"    Sequence: {message.Sequence}");
        Console.WriteLine($"    Size:     {message.SizeBytes} bytes");

        if (message.Timestamp != null)
        {
            Console.WriteLine($"    Time:     {message.Timestamp.ToDateTime():yyyy-MM-dd HH:mm:ss.fff}");
        }

        if (!string.IsNullOrEmpty(message.Consumer))
        {
            Console.WriteLine($"    Consumer: {message.Consumer}");
        }

        // Try to decode data as UTF-8 string
        if (message.Data != null && message.Data.Length > 0)
        {
            try
            {
                var dataStr = message.Data.ToStringUtf8();
                var preview = dataStr.Length > 100 ? dataStr.Substring(0, 100) + "..." : dataStr;
                Console.WriteLine($"    Data:     {preview}");
            }
            catch
            {
                Console.WriteLine($"    Data:     [binary, {message.Data.Length} bytes]");
            }
        }

        Console.WriteLine();
    }

    /// <summary>
    /// Run all examples
    /// </summary>
    public async Task RunAllExamples()
    {
        Console.WriteLine($"WebSocket Client Example - Connecting to {_baseUrl}");
        Console.WriteLine("=".PadRight(80, '='));
        Console.WriteLine();

        try
        {
            // Example 1: Stream from ephemeral consumer
            await StreamFromEphemeralConsumer("events.>", maxMessages: 5);

            // Example 2: Stream from durable consumer (will fail if consumer doesn't exist)
            // Uncomment if you have a durable consumer set up:
            // await StreamFromDurableConsumer("EVENTS", "my-durable-consumer", maxMessages: 5);

            // Example 3: Stream with timeout
            await StreamWithTimeout("events.test", timeoutSeconds: 10);

            Console.WriteLine("=".PadRight(80, '='));
            Console.WriteLine("✓ All examples completed successfully!");
        }
        catch (Exception ex)
        {
            Console.WriteLine($"✗ Error: {ex.Message}");
            Console.WriteLine($"  Make sure NatsHttpGateway is running at {_baseUrl}");
            Console.WriteLine($"  Make sure NATS is running and has messages to stream");
        }
    }
}

/// <summary>
/// Example program entry point for WebSocket client
/// </summary>
public class WebSocketProgram
{
    public static async Task Main(string[] args)
    {
        // Configuration priority: CLI arg > Environment variable > Default
        var baseUrl = args.Length > 0
            ? args[0]
            : Environment.GetEnvironmentVariable("NATS_GATEWAY_URL") ?? "http://localhost:5000";
        var client = new WebSocketClientExample(baseUrl);

        await client.RunAllExamples();

        Console.WriteLine();
        Console.WriteLine("Press any key to exit...");
        Console.ReadKey();
    }
}
